ONSdigital/rm-sample-service
# Upgrade to Java 21 spring boot 3

# What and why?

Upgrading Java and Spring Boot

- Upgrade to Java 21
- Upgrade to Spring Boot 3

# How to test?

- Install Java 21 SDK locally (tested with temurin-21.0.4)
- switch to Java 21 (add /Library/Java/JavaVirtualMachines/temurin-21.0.4/Contents/Home to local jenv versions and use jenv local 21)

- compile, run tests and build image locally
```
mvn clean install
```
- run the Postgres and PubSub emulators to allow the app start and connect
- run the sprint-boot app
```
docker run -d --name pubsub-emulator -p 8681:8681 -e PUBSUB_PROJECT1=local,test_subscription:test_subscription eu.gcr.io/ons-rasrmbs-management/pubsub-emulator
docker run --name postgres -p 5432:5432 -e POSTGRES_PASSWORD=postgres -e POSTGRES_USER=postgres -e POSTGRES_DB=ras -e POSTGRES_HOST_AUTH_METHOD=trust -d postgres:14

mvn spring-boot:run -Dspring-boot.run.arguments="--cloud.gcp.project-id=local --gcp.project=local"
java -jar target/sample.jar --cloud.gcp.project-id='local' --gcp.project='local' --PUBSUB_EMULATOR_HOST='http://localhost:8681'
```
The app should start and run but not terminate with a exit code (although you might see some unrelated `NOT_FOUND: Resource not found` subscription errors in the logs as the emulator is not set up correctly to run outside of the integration tests).

- check the Java 21 JVM starts (and not Java 17) and Spring Boot 3
- check there are no app errors and that the `sample` schema is created/updated and populated with data by the changeset releases
- kill the Maven run `spring-boot` app
- stop and delete the Postgres / Pub/Sub containers
```
docker stop postgres
docker rm postgres
docker stop pubsub-emulator
docker rm pubsub-emulator
```
- build a docker image locally
```
docker build  --no-cache -t eu.gcr.io/ons-rasrmbs-management/sample:java21 -f _infra/docker/Dockerfile .
```
- ensure the image builds successfully
- ideally test the image in a docker network if you maintain one locally
- deploy this branch image to your GCP development namespace
- ensure the app starts (Java 21 Spring Boot 3) and the acceptance tests pass
